### PR TITLE
[IMP] website_sale: product page improve carousel and grid

### PR DESCRIPTION
**Pull request by mano-odoo**
**DEV This PR still requires a functional change on the Image Ratio editor option to properly set the active option when the associated template is active** *at the moment it still displays "Default (1x1)" while the other template is active. (the class on the `<main>` is refreshed on reload and the we_editor looses context).*

This PR aims to review the single image, carousel and grid layout on the product page.

**Grid**
Previously the images were displayed in multiple columns while keeping there original ratio creating ugly gaps in the grid.

They feeds each column with image filling them in order and filling the column width with letting the height scale accordingly.

If the amount of image < amount of col set, it will show only a column if there is an image to put inside.

Limitation is when the user uploads a tall image with a small width which uneven the columns. 
But it comes down to user choice, assumption made that the images uploaded by the user should share more or less the same ratio in the grid.

Using a margin instead of a padding allows the hover effect in edition to better fit the image element. 
Note: This is still visually weird due to the overflow-hidden above, this is necessary for the ribbons and will be removed if the ribbons are reworked.

**Carousel** 
We add the possibility for the user to choose an aspect ratio to be used in the carousel, this allows better control on the height of the content to better fit the user needs.

Furthermore, we now remove the irrelevant options that was shown in the editor while the user is editing a page that contains a single image. This was misleading as changing a carousel inner option would impact the other pages while not providing any visual feedback on the current page the user is editing.

Additionally it adapts the carousel arrows the better fit the reviewed design and follow the button border radius set by the user. 

**Single image**
When using a single image, this image should fill the entire column width on the product page. This commit lets the image width fills it's container while letting the height adapt accordingly.

Relying on the width fixes the issue with the media_preview since it now scales evenly letting the play button always in the middle of its container.

task-3986938
part of task-3986921

| Before | After |
| -- | -- | 
|image | image| 
| Grid 3 col but 1 image v | Grid 3 col but 1 image v |
| image | image| 
| image | image | 
| Carousel portrait image before v | Carousel portrait image with new option v |
|image | image | 
| image | image |

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_sale/static/src/scss/website_sale.scss [hunk 1]
```diff
@@ -405,15 +405,12 @@ a.no-decoration {
...
         display: flex;
         align-items: center;
-        @include media-breakpoint-down(md) {
-            height: 300px;
-        }
-        @include media-breakpoint-down(sm) {
-            height: 200px;
+
+        &.o_carousel_not_single {
+            height: 25rem;
```
Comment (stefanorigano) [inline addons/website_sale/static/src/scss/website_sale.scss:413]: This is unfortunately "not enough" 😕

| |
|--------|
|image|

It's responsibility of the user to provide image of the "right size", yes,  but we should allow the use of portrait images here.
I propose:
```scss
&.o_carousel_not_single {
    height: 25rem;
  
    .product_detail_img {
        width: auto !important; // overrides .w-100
        max-height: 100%;
    }
}    
```

| |
|--------|
| image | 


...even better, we could keep the behavior described before while allowing the user to set a specific aspect-ratio  : 
```scss
 &.o_carousel_not_single {
            &.o_ratio-1x1 {
                aspect-ratio: 1 / 1;
            }

            &.o_ratio-4x3 {
                aspect-ratio: 4 / 3;
            }

            &.o_ratio-16x9 {
                aspect-ratio: 16 / 9;
            }

            &.o_ratio-21x9 {
                aspect-ratio: 21 / 9;
            }

            &.o_ratio-4x5 {
                aspect-ratio: 4 / 5;
            }
     
            .product_detail_img {
                width: auto !important; // overrides .w-100
                max-height: 100%;
            }
        }
  
    .product_detail_img {
        width: auto !important;
    }
}    
```

`addons/website_sale/views/snippets/snippets.xml`
```xml
<we-select string="Layout" data-name="o_wsale_image_layout" data-no-preview="true" data-reload="/">
    <we-button data-set-image-layout="carousel" data-name="o_wsale_carousel_opt">Carousel</we-button>
    <we-button data-set-image-layout="grid">Grid</we-button>
</we-select>
<we-select string="Images Ratio" data-name="o_wsale_carousel_height" data-apply-to=".o_carousel_product_outer" data-dependencies="o_wsale_carousel_opt" class="o_we_sublevel_1">
    <we-button data-select-class="o_ratio-4x5">Portait (4/5)</we-button>
    <we-button data-select-class="o_ratio-1x1">Square (1/1)</we-button>
    <we-button data-select-class="o_ratio-4x3">Landscape (4/3)</we-button>
    <we-button data-select-class="o_ratio-16x9">Wide (16/9)</we-button>
    <we-button data-select-class="o_ratio-21x9">Wider (21/9)</we-button>
</we-select>
```

|  |
|--------|
| image |

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Not bad @mano-odoo 👍 
I knew that it was possible to address this, well done.

Some minor remarks + a suggestion to address  the most prominent issue remaining.
Also, if you could fix the indicators aspect-ratio it would be great 😉
|  |
|--------|
| image | 

Nitpicking:
- 18d30616916a50b054581d88fe9d140a5bd178bb commit description: rather than "Part of website_sale product redesign (task-3986921)" at the beginning, use this  formula at the end:
```
task-3986938
part-of task-3986921
```

File: addons/website_sale/static/src/xml/website_sale_image_viewer.xml [hunk 2]
```diff
@@ -24,7 +24,10 @@
                                 <t t-foreach="images" t-as="image" t-key="image.thumbnailSrc">
                                     <li t-attf-class="align-top position-relative px-1 pb-1 {{image === selectedImage ? 'active' : ''}}" t-on-click="() => this.selectedImage = image">
                                         <div>
-                                            <img t-att-src="image.thumbnailSrc" t-attf-class="img o_wsale_image_viewer_thumbnail {{image === selectedImage ? 'active' : ''}}" t-att-alt="props.title" loading="lazy"/>
+                                            <img t-att-src="image.thumbnailSrc"
+                                                t-attf-class="img o_wsale_image_viewer_thumbnail object-fit-cover {{image === selectedImage ? 'active' : ''}}"
```
Comment (stefanorigano) [inline addons/website_sale/static/src/xml/website_sale_image_viewer.xml:28]: Let's use `contain` 😉

|  |
|--------|
| image | 


```suggestion
                                                t-attf-class="img o_wsale_image_viewer_thumbnail bg-body object-fit-contain {{image === selectedImage ? 'active' : ''}}"
```

File: addons/website_sale/views/snippets/snippets.xml [hunk 3]
```diff
@@ -280,6 +280,32 @@
                 <we-button data-set-image-layout="carousel">Carousel</we-button>
                 <we-button data-set-image-layout="grid">Grid</we-button>
             </we-select>
+            <we-select string="Images Ratio" data-name="o_wsale_image_ratio" class="o_we_sublevel_1" data-reload="/">
+                <we-button data-select-class=""
+                           data-customize-website-views="">
+                           None
+                </we-button>
+                <we-button data-select-class="o_carousel_ratio_1x1"
```
Comment (stefanorigano) [inline addons/website_sale/views/snippets/snippets.xml:288]: Prefix classes according to their module scope 😉

```suggestion
                <we-button data-select-class="o_wsale_carousel_ratio_1x1"
```

**Review by stefanorigano — APPROVED** [review]
Hello @mano-odoo 👋
One remark and one nitpicking comment for you, all the rest LGTM.
Since this will go to tech. review anyway, I'll not block this further 👍 

Thanks for your work here, good job!

File: addons/website_sale/static/src/xml/website_sale_image_viewer.xml (hunk 2 above)
Comment (stefanorigano) [reply addons/website_sale/static/src/xml/website_sale_image_viewer.xml:28]: .. and don't forget the mobile adaptation we discussed about ;)

**Comment by dalaOdoo** [comment]
thumbnails on the left are broken 
image

**Comment by mano-odoo** [comment]
Fixed thumbnails on the left ✅ 
Fixed "100_pc" thumbnail and image misaligned ✅ 
Fixed the missing dataset traceback ✅

**Review by chevalierv — CHANGES_REQUESTED** [review]
Hello @mano-odoo @stefanorigano 

Thanks for improving our eCommerce! The product images look neater than never :star_struck: 

I've pushed small fixups of nitpicking and fixed the bug with the image ratio. Feel free to squash it whenever you're ready :smiley:

FYI the arrow on the carousel changes position when using a custom ration (default 1x1 works as expected). I don't know if that's expected or not.

Have a nice day! :sun_with_face:

image

**Comment by mano-odoo** [comment]
@chevalierv 
> FYI the arrow on the carousel changes position when using a custom ration (default 1x1 works as expected). I don't know if that's expected or not.

[This was due to the t-attf-class](https://github.com/odoo/odoo/compare/183a19369645dcbf9fbcecce8f296bcdbc4deafa..cad8fa759bdd76ad231cf8f3769014078f3f6fde) in the xpath, I changed the attribute to a `class` and kept the xpath with the `t-attf-` attribute, thus creating an new attribute and taking priority over the `class` one, erasing all the necessary classes on `o_carousel_product_outer`😵‍💫 

For the image_viewer previews I forced the width to 128px, so if your image is too small it should still fill the container 😄

File: addons/website_sale/static/src/xml/website_sale_image_viewer.xml [hunk 4]
```diff
@@ -24,19 +24,50 @@
...
+                            >
+                                <i
+                                    class="oi oi-chevron-right border bg-white text-900"
+                                    role="img"
+                                    aria-label="Next"
+                                    title="Next"
+                                />
+                            </a>
                         </div>
+
```
Comment (chevalierv) [inline addons/website_sale/static/src/xml/website_sale_image_viewer.xml:70]: nitpicking: why do you add this new line?
```suggestion
```

**Review by chevalierv — APPROVED** [review]
Hello @mano-odoo 

Thanks for your work on this! LGTM now :+1: 

@robodoo delegate+

Have a nice day!

**Comment by mano-odoo** [comment]
@robodoo r+ rebase-ff

**Comment by mano-odoo** [comment]
@robodoo r+

**Pull request closed without merging**